crates/nu-protocol/src/config/reedline.rs

```diff
@@ -47,18 +47,48 @@ impl FromStr for NuCursorShape {
         "blink_block" => Ok(NuCursorShape::BlinkBlock),
         "blink_underscore" => Ok(NuCursorShape::BlinkUnderscore),
         "inherit" => Ok(NuCursorShape::Inherit),
-        _ => Err("expected either 'line', 'block', 'underscore', 'blink_line', 'blink_block', 'blink_underscore' or 'inherit'"),
+        _ => Err("'line', 'block', 'underscore', 'blink_line', 'blink_block', 'blink_underscore' or 'inherit'"),
         }
     }
 }
 
+impl UpdateFromValue for NuCursorShape {
+    fn update(&mut self, value: &Value, path: &mut ConfigPath, errors: &mut ConfigErrors) {
+        config_update_string_enum(self, value, path, errors)
+    }
+}
+
 #[derive(Clone, Copy, Debug, Default, IntoValue, PartialEq, Eq, Serialize, Deserialize)]
 pub struct CursorShapeConfig {
     pub emacs: NuCursorShape,
     pub vi_insert: NuCursorShape,
     pub vi_normal: NuCursorShape,
 }
 
+impl UpdateFromValue for CursorShapeConfig {
+    fn update<'a>(
+        &mut self,
+        value: &'a Value,
+        path: &mut ConfigPath<'a>,
+        errors: &mut ConfigErrors,
+    ) {
+        let Value::Record { val: record, .. } = value else {
+            errors.type_mismatch(path, Type::record(), value);
+            return;
+        };
+
+        for (col, val) in record.iter() {
+            let path = &mut path.push(col);
+            match col.as_str() {
+                "vi_insert" => self.vi_insert.update(val, path, errors),
+                "vi_normal" => self.vi_normal.update(val, path, errors),
+                "emacs" => self.emacs.update(val, path, errors),
+                _ => errors.unknown_option(path, val),
```

Comment:
Will this new pattern still remove the invalid options from the backing `Record` or will they stick around and re-raise?

Reply:
The new config is turned back into a value which is then used to overwrite the old env value. So invalid options are discarded in the process.